src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/RemoveEmptyFinalizers.cs

```diff
@@ -33,31 +35,45 @@ public override void Initialize(AnalysisContext analysisContext)
             analysisContext.EnableConcurrentExecution();
             analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 
-            analysisContext.RegisterCodeBlockAction(codeBlockContext =>
+            analysisContext.RegisterOperationBlockStartAction(obsac =>
             {
-                if (codeBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                if (!obsac.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsConditionalAttribute, out var conditionalAttributeType) ||
+                    obsac.OperationBlocks.Length != 1 ||
+                    !(obsac.OperationBlocks[0] is IBlockOperation blockOperation) ||
+                    !(obsac.OwningSymbol is IMethodSymbol methodSymbol) ||
+                    !methodSymbol.IsDestructor())
                 {
                     return;
                 }
 
-                var methodSymbol = (IMethodSymbol)codeBlockContext.OwningSymbol;
-                if (!methodSymbol.IsDestructor())
+                const int IS_EMPTY = 0;
+                const int IS_NOT_EMPTY = 1;
+                var state = IS_EMPTY;
+
+                obsac.RegisterOperationAction(oac =>
                 {
-                    return;
-                }
+                    if (Interlocked.CompareExchange(ref state, IS_EMPTY, IS_EMPTY) == IS_EMPTY &&
```

Comment:
I am pretty beginner with `Interlocked` use so there might be better solutions to have this flag done in a thread-safe manner. Note that I have tried to avoid running the lambda if we have already found any valid statement.